src/Symfony/Component/Messenger/Handler/HandlersLocator.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Handler;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class HandlersLocator implements HandlersLocatorInterface
+{
+    /**
+     * Maps a message to a list of handlers.
+     */
+    private $handlers;
+
+    /**
+     * @param callable[][] $handlers
+     */
+    public function __construct(array $handlers)
+    {
+        $this->handlers = $handlers;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getHandlers(string $name): iterable
+    {
+        $seen = array();
+
+        foreach (self::listTypes($name) as $type) {
+            foreach ($this->handlers[$type] ?? array() as $handler) {
+                if (!\in_array($handler, $seen, true)) {
+                    yield $seen[] = $handler;
+                }
+            }
+        }
+    }
+
+    /**
+     * @internal
+     */
+    public static function listTypes(string $class): array
+    {
+        return array($class => $class)
+            + class_parents($class)
```

Comment:
It does not seem to be fixed that name must be a class name. So calling these methods is not safe.